src/box/lua/schema.lua

```diff
@@ -3593,30 +3593,84 @@ local function role_check_grant_revoke_of_sys_priv(priv)
     end
 end
 
-box.schema.role.grant = function(user_name, ...)
+local function box_schema_role_grant_do(user_name, ...)
     local uid = role_resolve(user_name)
     if uid == nil then
         box.error(box.error.NO_SUCH_ROLE, user_name)
     end
     role_check_grant_revoke_of_sys_priv(...)
     return grant(uid, user_name, ...)
 end
-box.schema.role.revoke = function(user_name, ...)
+
+local function box_schema_role_revoke_do(user_name, ...)
     local uid = role_resolve(user_name)
     if uid == nil then
         box.error(box.error.NO_SUCH_ROLE, user_name)
     end
     role_check_grant_revoke_of_sys_priv(...)
     return revoke(uid, user_name, ...)
 end
-box.schema.role.info = function(role_name)
+
+local function box_schema_role_info_do(role_name)
     local rid = role_resolve(role_name)
     if rid == nil then
         box.error(box.error.NO_SUCH_ROLE, role_name)
     end
     return info(rid)
 end
 
+-- Wrap a function into transaction if none is active.
+local function atomic_wrapper(func)
+    return function(...)
+        -- No reason to start a transaction of one is active already.
+        if box.is_in_txn() then
+            return func(...)
+        else
+            return box.atomic(func, ...)
+        end
+    end
+end
+
+box.schema.space.create = atomic_wrapper(box_schema_space_create_do)
+box.schema.space.format = atomic_wrapper(box_schema_space_format_do)
+box.schema.space.upgrade = box_schema_space_upgrade_do
+box.schema.space.drop = atomic_wrapper(box_schema_space_drop_do)
+box.schema.space.rename = atomic_wrapper(box_schema_space_rename_do)
+box.schema.space.alter = atomic_wrapper(box_schema_space_alter_do)
+
+box.schema.index.create = atomic_wrapper(box_schema_index_create_do)
+box.schema.index.drop = atomic_wrapper(box_schema_index_drop_do)
+box.schema.index.rename = atomic_wrapper(box_schema_index_rename_do)
+box.schema.index.alter = box_schema_index_alter_do
```

Comment:
Why is this function not wrapped in atomic? AFAIU it may perform a few statements.

Reply:
It's not possible to make it atomic for now. The function can perform index rebuild/format check, which [are](https://github.com/tarantool/tarantool/commit/f266559b167b4643c377724eebde9651877d6cc9#diff-4783b9090bf7018ce3825d26935a097236f7caf5d87dc2fb0e7fa4e1d3c3f86bR1041) [forced](https://github.com/tarantool/tarantool/commit/f266559b167b4643c377724eebde9651877d6cc9#diff-4783b9090bf7018ce3825d26935a097236f7caf5d87dc2fb0e7fa4e1d3c3f86bR893) to be the first statements in a transaction in order to [prevent](https://github.com/tarantool/tarantool/pull/6228#issuecomment-891620195) other transactions from seeing partial schema changes.

At the same time it can [delete](https://github.com/tarantool/tarantool/blob/52884400993d1bca1c490e3a739f5528ce06ed3b/src/box/lua/schema.lua#L1366) an item from the `_space_sequence` table, which will be the first transaction statement. So the index rebuild/format check will fail. And we can't drop the sequence after altering the index because the alter can change the type of the field we have the existing sequence for, so the attached to the space sequence will become [invalid](https://github.com/tarantool/tarantool/blob/52884400993d1bca1c490e3a739f5528ce06ed3b/src/box/alter.cc#L164) for its key def.

So the function in guaranteed to fail in case of specific input if it's wrapped into a transaction. The only way to fix this I see is to implement the [transactional DDL](https://github.com/tarantool/tarantool/issues/600).

Reply:
Why do we have to delete an item from the `_space_sequence` table *before* updating an item in the `_index` space? Why can't we do that after we're done building the index?

Reply:
Dropped the sequence check from the `_index:on_replace` trigger and moved the sequence unassignment after the index alter.

This has a caveat though: uncareful manual altering the index can silently put the space into an inconsistent condition: the next insert into the space can either fail or silently succeed whereas it's not supposed operation for the space. Example:

```lua
box.cfg{}
s = box.schema.space.create('s')
pk = s:create_index('pk', {parts = {{1, 'unsigned'}}, sequence = true})

-- Manual altering the index without dropping the sequence.
box.space._index:update({s.id, pk.id}, {{'=', 6, {{0, 'string'}}}})

-- This will fail: passed integer (from sequence) in string field.
s:insert({box.NULL})

-- Manual altering the index without dropping the sequence again.
box.space._index:update({s.id, pk.id}, {{'=', 6, {{0, 'scalar'}}}})

-- This will succeed: passed integer (from sequence) in scalar field.
-- Though it shouldn't have been done at all: one should not pass
-- NULL in the field which is indexed by the primary key.
s:insert({box.NULL})
```